pulp_gem/specs.py

```diff
@@ -10,10 +12,80 @@
 import rubymarshal.reader
 
 
+NAME_REGEX = re.compile(r"[\w\.-]+")
+VERSION_REGEX = re.compile(r"\d+(?:\.\d+)*")
+PRERELEASE_VERSION_REGEX = NAME_REGEX
+
 # Natural key.
 Key = namedtuple("Key", ("name", "version"))
 
 
+async def read_versions(relative_path):
+    # File starts with:
+    #   created_at: <timestamp>
+    #   ---
+    async with aiofiles.open(relative_path, mode="r") as fp:
+        results = {}
+        preamble = True
+        async for line in fp:
+            line = line.strip()
+            if line == "---":
+                preamble = False
+                continue
+            if preamble:
+                continue
+            name, versions, md5_sum = line.split(" ", maxsplit=2)
+            versions = versions.split(",")
+            entry = results.get(name) or ([], "")
+            results[name] = (entry[0] + versions, md5_sum)
+    for name, (versions, md5_sum) in results.items():
+        # Sanitize name
+        if not NAME_REGEX.match(name):
+            raise ValueError(f"Invalid gem name: {name}")
```

Comment:
Have you ran this regex through all the gem names returned off of rubygems?

Reply:
Yes.
```python
In [1]: from pulp_gem.specs import NAME_REGEX

In [2]: import requests

In [3]: response = requests.get("https://rubygems.org/names")

In [4]: [name for name in response.text.split("\n") if not NAME_REGEX.match(name)]
Out[4]: ['']
```